### ISSUE TITLE
Supporting Contract Signatures for undeployed contracts

### ISSUE THREAD

**Issue by simon-jentzsch** [matter-labs/zksync#256]
Currently L1-Accounts as Multisig are supported by checking the Contract Signature with EIP-1271 on the server and an L1 authorization when setting the pub key using `setAuthPubkeyHash`.

But this requires a Contract to be deployed on L1 first. As a feature request, we would love to support signature verification without deploying the contract.
This means as part of the signature we could pass the code of the deployment-transaction for L1. In order to verify we need to 
- create a EVM
- deploy the tx-code
- call `isValidSignature` for the account.

To optimize we may cache the code or even better offer precompiled modules handling the logic directly based on the tx-data. ( For most Multisigs this could easily be implemented as Rust-Code)
But since there are 2 places we need to verify signatures (the Rust-Code and in the SolidityContract), this would also need to implemented in the `verifyChangePubkeySignature` function.

We would love to support you, since this is an important feature, which would bring down deployment costs and would allow us to use multisigs in zksync without deploying them.

**Comment by dvush** [comment]
> As a feature request, we would love to support signature verification without deploying the contract.

Great! We are implementing this feature right now.

Here how it would work:

You generate zkSync private key and compute your account address using `CREATE2` spec where `salt` is hashed zkSync pubkey_hash. Then you can set this pubkey_hash as signing key.

Prototype https://github.com/matter-labs/create2-example

**Comment by simon-jentzsch** [comment]
This is awesome! I will test it out and let you know in case of any issues!

**Comment by simon-jentzsch** [comment]
Yes, just tested it and this would indeed work for us! Tanks a lot.
There is only one issue:
Currently the salt is calculated from the signer key and the used to calc the address of the account.

Since we are deploying the gnosis safe, this would not work, becasue the safe will hash the initiator-data to the salt, which means in order to verify, you would have to also add this.
Do you think we can find a generic solution to add additional data before hashing?

bytes32 salt = keccak256(abi.encodePacked(keccak256(initializer), saltNonce));
https://github.com/gnosis/safe-contracts/blob/development/contracts/proxies/GnosisSafeProxyFactory.sol

**Comment by simon-jentzsch** [comment]
I just created a small PR to simply switch the order: https://github.com/matter-labs/zksync/pull/259

**Comment by dvush** [comment]
Ok, great that you reported it now, since we did not upgrade our contracts yet.

**Comment by simon-jentzsch** [comment]
By the way I made a small change in the pull-request and actually tested it with the gnosis safe. (I added padding to 32bytes, because this is how all the gnosis safes are deployed) Now It works! 
I generated a seed and from this the pub_key_hash. Deployed the GnosisSafeFactory, (which is the creatorAddress in this case) and used the codehas of the Proxy. Based on this I calculated the my account-address. and submitted a ChnagePubKey-tx, which was successfully executed and verified.
Also transfer- and withdraw worked great.
